packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,501 @@
+    }
+
+    /// @dev Sets the precompiles, proxies, and the implementation accounts to be `vm.dumpState`
+    ///      to generate a L2 genesis alloc.
+    /// @notice The alloc object is sorted numerically by address.
+    function run() public {
+        _setPrecompiles();
+        _setProxies();
+        _setImplementations();
+
+        /// Reset so its not included state dump
+        vm.etch(address(cfg), "");
+
+        vm.dumpState(outfile);
+        _sortJsonByKeys(outfile);
+    }
+
+    /// @notice Give all of the precompiles 1 wei so that they are
+    ///         not considered empty accounts.
+    function _setPrecompiles() internal {
+        for (uint256 i; i < PRECOMPILE_COUNT; i++) {
+            vm.deal(address(uint160(i)), 1);
+        }
+    }
+
+    /// @dev Set up the accounts that correspond to the predeploys.
+    ///      The Proxy bytecode should be set. All proxied predeploys should have
+    ///      the 1967 admin slot set to the ProxyAdmin predeploy. All defined predeploys
+    ///      should have their implementations set.
+    function _setProxies() internal {
+        bytes memory code = vm.getDeployedCode("Proxy.sol:Proxy");
+        uint160 prefix = uint160(0x420) << 148;
+        for (uint256 i = 0; i < PROXY_COUNT; i++) {
+            address addr = address(prefix | uint160(i));
+            if (_notProxied(addr)) {
+                continue;
+            }
+
+            console.log("Seting proxy deployed bytecode at: %s", addr);
+            vm.etch(addr, code);
+            vm.store(addr, PROXY_ADMIN_ADDRESS, bytes32(uint256(uint160(Predeploys.PROXY_ADMIN))));
+
+            if (_isDefinedPredeploy(addr)) {
+                address implementation = _predeployToCodeNamespace(addr);
+                console.log(
+                    "Setting proxy %s implementation: %s",
+                    addr, implementation
+                );
+                vm.store(addr, PROXY_IMPLEMENTATION_ADDRESS, bytes32(uint256(uint160(implementation))));
+            }
+        }
+    }
+
+    /// @notice LEGACY_ERC20_ETH is not being predeployed since it's been deprecated.
+    /// @dev Sets all the implementations for the predeploy proxies. For contracts without proxies,
+    ///      sets the deployed bytecode at their expected predeploy address.
+    function _setImplementations() internal {
+        _setLegacyMessagePasser();
+        _setDeployerWhitelist();
+        _setWETH9();
+        _setL2StandardBridge();
+        _setL2CrossDomainMessenger();
+        _setSequencerFeeVault();
+        _setOptimismMintableERC20Factory();
+        _setL1BlockNumber();
+        _setGasPriceOracle();
+        _setGovernanceToken();
+        _setL1Block();
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    function _setLegacyMessagePasser() internal {
+        address impl = _predeployToCodeNamespace(Predeploys.LEGACY_MESSAGE_PASSER);
+        console.log(
```

Comment:
We could wrap this in an internal fn that takes name + address to dedupe code

Reply:
Good call! Deduped a lot of code!